kvserver: add TestCreateManyUnappliedProbes

tbg opened the pull request:
This is the test used for https://github.com/cockroachdb/cockroach/pull/102953.

This PR also makes modest progress on #75729, not by making log application
stand-alone, but by making it somewhat less convoluted to manufacture log
entries programmatically. It would be desirable to be able to test the
replication layer in CRDB in the same way that Raft allows via the
InteractionEnv[^1].

[^1]: https://github.com/etcd-io/raft/blob/6bf4f7fe3122b064e0a0d76314298dca6f379fc7/interaction_test.go

Epic: none
Release note: none

tbg committed:
kvserver: move some logging in anticipation of a refactor

Epic: none

Release note: None

tbg committed:
kvserver: begin extracting raftCmdToPayload

Epic: none
Release note: None

tbg committed:
kvserver: extract raftCmdToPayload

Epic: none
Release note: None

tbg committed:
kvserver: move and rename makeIDKey()

Epic: none
Release note: none

tbg committed:
kvserver: move change replicas validation out of raftCmdToPayload

Epic: none
Release note: none

tbg committed:
kvserver: move raftCmdToPayload to raftlog

Epic: none
Release note: none

tbg committed:
kvserver: add TestCreateManyUnappliedProbes

This is the test used for https://github.com/cockroachdb/cockroach/pull/102953.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/104401)
<!-- Reviewable:end -->

pavelkalinnikov commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -500,8 +404,8 @@ func (r *Replica) propose(
 	const largeProposalEventThresholdBytes = 2 << 19 // 512kb
-	if cmdLen > largeProposalEventThresholdBytes {
-		log.Eventf(p.ctx, "proposal is large: %s", humanizeutil.IBytes(int64(cmdLen)))
+	if len(p.encodedCommand) > largeProposalEventThresholdBytes {
+		log.Eventf(p.ctx, "proposal is large: %s", humanizeutil.IBytes(int64(len(p.encodedCommand))))
```
```suggestion
	if ln := len(p.encodedCommand); ln > largeProposalEventThresholdBytes {
		log.Eventf(p.ctx, "proposal is large: %s", humanizeutil.IBytes(int64(ln)))
```

pavelkalinnikov commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -500,8 +404,8 @@ func (r *Replica) propose(
 	const largeProposalEventThresholdBytes = 2 << 19 // 512kb
-	if cmdLen > largeProposalEventThresholdBytes {
-		log.Eventf(p.ctx, "proposal is large: %s", humanizeutil.IBytes(int64(cmdLen)))
+	if len(p.encodedCommand) > largeProposalEventThresholdBytes {
```
This has switched from `p.command.Size()` to `len(p.encodedCommand)`. Is this ok? Is this the same size?

tbg replied:
It's not the same size, but it's comparable. The threshold is pretty arbitrary too.

pavelkalinnikov commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -537,6 +428,115 @@ func (r *Replica) propose(
 }
 
+func raftCmdToPayload(
+	_ context.Context,
```
Unused. Remove?

pavelkalinnikov commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -380,7 +379,7 @@ func (r *Replica) propose(
 	}
 
-	data, err := raftCmdToPayload(ctx, p.command, p.idKey)
+	data, err := raftlog.RaftCmdToPayload(ctx, p.command, p.idKey)
```
The name is redundant. Consider `raftlog.CommandToPayload` or `raftlog.EncodeCommand`.

pavelkalinnikov commented on pkg/kv/kvserver/client_manual_proposal_test.go:
```diff
@@ -0,0 +1,270 @@
+// Copyright 2022 The Cockroach Authors.
```
```suggestion
// Copyright 2023 The Cockroach Authors.
```

pavelkalinnikov commented on pkg/kv/kvserver/client_manual_proposal_test.go:
```diff
@@ -0,0 +1,270 @@
+			}
+			res, err := cmd.EvalRW(ctx, b, args, resp)
+			require.NoError(t, err)
+			// TODO: there's more stuff in evaluateProposal that would need to be lifted
```
Is the TODO for this PR, or later? Make it TODO(name/issue)?

pavelkalinnikov commented on pkg/kv/kvserver/client_manual_proposal_test.go:
```diff
@@ -0,0 +1,270 @@
+
+	ctx := context.Background()
+
+	// NB: I used this as follows:
```
Some clarification on what this does would be helpful. As I understood it:

You started a real CRDB node, and killed it after 10s. It populated a correct database instance on disk. Then you started this unit-test to append a bunch of log entries to the Raft log of range 10. Then you started the node again to see how it applies the new entries.

tbg committed:
raftlog: rename EncodeRaftCommand -> EncodeCommandBytes

That way, we can then rename RaftCmdToPayload to EncodeCommand without creating
confusion.

EncodeRaftCommand is only used in unquiesce (and testing).

Epic: none
Release note: None

tbg commented:
Hold off on review, I want to first rebase on top of the replication admission control change.

tbg commented:
I had to basically re-do this PR as part of the rebase. I took the opportunity to rearrange the commits somewhat, and to bake in the renames we decided on earlier. Even though the structure follows very closely to the original work, please give it an honest review. (I self-reviewed already but one is usually blind to one's own slip-ups).
Thanks!

pavelkalinnikov commented on pkg/kv/kvserver/replica_raft.go:
```diff
@@ -383,154 +383,170 @@ func (r *Replica) propose(
-		preLen = raftlog.RaftCommandPrefixLen
-	}
-
-	raftAdmissionMeta := &kvflowcontrolpb.RaftAdmissionMeta{}
```
Did we use the facts that this is non-nil? Seems like no, so we're good.

pavelkalinnikov commented on pkg/kv/kvserver/client_manual_proposal_test.go:
```diff
@@ -0,0 +1,272 @@
+
+			// End of evaluation. Start of "proposing".
+
+			// TODO: the "requires consensus" logic is not reusable, make it so:
```
Is there an issue? More generally, is this test going to be unskipped at some point in some other form? Do we need to avoid this test being dead code?

If this test is intended as a demo, link it in the reverse direction from the issue that it serves? So that it's not forgotten.

tbg replied:
Filed https://github.com/cockroachdb/cockroach/issues/105177 and cross-referring from the comments to that issue and back.
